### PR TITLE
Upload/Fetch from URL/Use OCDS Data Source - back

### PR DESCRIPTION
**Pull request by VDigitall**
Component of backend allowing users to input JSON

The input JSON can be provided in a few ways:
* Upload a file
* Provide the URL of a file
* Via the OCDS data registry

Resolve issue https://github.com/open-contracting/spoonbill/issues/3

### CODE REVIEW

File: core/tasks.py [hunk 1]
```diff
@@ -1,53 +1,239 @@
...
-        f"validate_data_{upload_id}",
-        {
-            "type": "task.validate",
-            "message": {"upload_id": upload_id, "is_valid": True},
-        },
+        f"validate_data_{datasource.id}",
+        {"type": "task.validate", "datasource": serializer.to_representation(instance=datasource)},
+    )
+
+    logger.info("Start validation for %s file" % object_id)
```
Comment (yshalenyk) [inline core/tasks.py:45]: maybe debug would do

File: core/models.py [hunk 2]
```diff
@@ -26,9 +26,16 @@ class Upload(models.Model):
...
     """
 
+    QUEUED_VALIDATION = "queued.validation"
+    VALIDATION = "validation"
+    STATUS_CHOICES = [
+        (QUEUED_VALIDATION, _("Queued validation")),
+        (VALIDATION, _("Validation")),
+    ]
     id = models.UUIDField(primary_key=True, editable=False, default=uuid.uuid4)
     filename = models.CharField(max_length=64)
```
Comment (yshalenyk) [inline core/models.py:36]: is it enough to hold whole path?

File: core/tasks.py [hunk 3]
```diff
@@ -1,53 +1,239 @@
...
+                    "type": "task.download_data_source",
+                    "datasource": serializer.to_representation(instance=datasource),
+                },
+            )
+            return
+        size = int(r.headers.get("Content-Length", 0))
+        downloaded = 0
+        chunk_size = 10240
+        os.mkdir(f"{settings.UPLOAD_PATH_PREFIX}{object_id}")
+        path = f"{settings.UPLOAD_PATH_PREFIX}{object_id}/{datasource.filename}"
```
Comment (yshalenyk) [inline core/tasks.py:137]: I think its betten to use django storage api to work with files

File: core/tasks.py [hunk 4]
```diff
@@ -1,53 +1,239 @@
...
+
+        channel_layer = get_channel_layer()
+        async_to_sync(channel_layer.group_send)(
+            f"validate_data_{object_id}",
+            {
+                "type": "task.download_data_source",
+                "datasource": serializer.to_representation(instance=datasource),
+            },
+        )
+        logger.info(
```
Comment (yshalenyk) [inline core/tasks.py:107]: probably debug

File: core/tasks.py (hunk 3 above)
Comment (yshalenyk) [reply core/tasks.py:137]: also, i think we should escape `datasource.filename`

File: core/models.py (hunk 2 above)
Comment (VDigitall) [reply core/models.py:36]: in this field store only filename like `dataset.json`

File: core/tasks.py (hunk 4 above)
Comment (VDigitall) [reply core/tasks.py:107]: ok

File: core/tasks.py (hunk 1 above)
Comment (VDigitall) [reply core/tasks.py:45]: ok

File: core/tasks.py (hunk 3 above)
Comment (VDigitall) [reply core/tasks.py:137]: I think better would be to rewrite the filename using some hash func like uuid4

**Comment by yshalenyk** [comment]
Looks good. But docker image starting to massive with all this dependencies, so I really want to try two stage build. Maybe lets wait a bit as @pontostroy try this out before merging.

**Pull request merged**